src/analysis/processing/qgsalgorithmfillnodata.h

```diff
@@ -0,0 +1,78 @@
+                         qgsalgorithmfillnodata.h
+                         ---------------------
+    begin                : January 2020
+    copyright            : (C) 2020 by Clemens Raffler
+    email                : clemens dot raffler at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#ifndef QGSALGORITHMFILLNODATA_H
+#define QGSALGORITHMFILLNODATA_H
+
+#define SIP_NO_FILE
+
+#include "qgis_sip.h"
+#include "qgsprocessingalgorithm.h"
+#include "qgsapplication.h"
+
+///@cond PRIVATE
+
+/**
+ * Fill NoData algorithm:
+ * This is an algorithm that fills the NoDataValues in a raster dataset
+ * based on a dynamic input parameter
+ */
+class QgsFillNoDataAlgorithm : public QgsProcessingAlgorithm
+{
+  public:
+
+    QgsFillNoDataAlgorithm() = default;
+    void initAlgorithm( const QVariantMap &configuration = QVariantMap() ) override;
+    QIcon icon() const override { return QgsApplication::getThemeIcon( QStringLiteral( "/algorithms/mAlgorithmFillNoData.svg" ) ); }
+    QString svgIconPath() const override { return QgsApplication::iconPath( QStringLiteral( "/algorithms/mAlgorithmFillNoData.svg" ) ); }
+    QString name() const override;
+    QString displayName() const override;
+    QStringList tags() const override;
+    QString group() const override;
+    QString groupId() const override;
+    QString shortHelpString() const override;
+    QgsFillNoDataAlgorithm *createInstance() const override SIP_FACTORY;
+
+  protected:
+    bool prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback ) override;
+    QVariantMap processAlgorithm( const QVariantMap &parameters,
+                                  QgsProcessingContext &context,
+                                  QgsProcessingFeedback *feedback ) override;
+
+  private:
+    QgsRasterLayer *mInputRaster;
+    double mFillValue = 0;
+    int mBand;
+    std::unique_ptr< QgsRasterInterface > mInterface;
+    QgsRectangle mExtent;
+    QgsCoordinateReferenceSystem mCrs;
+    int mLayerWidth;
+    int mLayerHeight;
+    int mNbCellsXProvider = 0;
+    int mNbCellsYProvider = 0;
+    double mInputNoDataValue;
+    QgsRasterDataProvider *mDestinationRasterProvider;
+
+    bool mDynamicFillValue = false;
+    QgsProperty mDynamicFillValueProperty;
```

Comment:
not needed

Reply:
First I thought about implementing a dynamic parameter. But I saw too little use cases for it - although it would be interesting to have something like `$cellmidpoint` available in the expression variables for some dynamic geometry calculations per cell.

Reply:
True -- but I think it would be of limited use in this particular algorithm. A "create/update raster using expression" algorithm sounds very interesting though!